docs/proposals/fua-restrict-feature/fua-restrict-feature.md

```diff
@@ -0,0 +1,127 @@
+AVP or the Filter-Id AVP in the Credit-Control-Answer message.
+
+```
+  [Final-Unit-Indication]
+    {Final-Unit-Action}
+    *[Restriction-Filter-Rule]
+    *[Filter-Id]
+    [Redirect-Server]
+      {Redirect-Address-Type}
+      {Redirect-Server-Address}
+```
+
+Restriction-Filter-Rule: of type IPFilterRule (FlowDescription). This will
+need to be translated to a PolicyRule.
+Filter-Id: Provides IP packet filter identifiers (In our case, It will be a
+list of PolicyRule identifiers)
+
+It is important to note that QoS can be enabled only when using Filter-Id AVP.
+Restriction-Filter-Rule AVP would only provide allowed traffic filters.
+
+This implementation will not provide Restriction-Filter-Rule as Filter-id
+provides support of all the required use-cases. Nevertheless, 
+Restriction-Filter-Rule could be translated to restriction dynamic rules with
+lower priority and apply them similar to static rules. 
+LocalEnforcer will ensure priority in this case.
+
+As we would like to use any static rule when restriction is enabled, there will
+no change to PolicyRules definition.
+
+## How We Will Change Magma
+
+**PolicyRule**
+
+Policy Object definition remains the same as the current definition. At the
+opposite of a policy rule used in the FUA-Redirect case, there is no need to
+introduce any additional field for FUA-Restrict.
+We will be able to apply any static rule when restricting access.
+
+
+**Session Proxy**
+
+We will limit this implementation to Filter-Id AVP only as It covers the desired
+use-cases. The AVP will reference a PolicyRule Identifier.
+
+```
+  [Final-Unit-Indication]
+    {Final-Unit-Action}
+    *[Filter-Id]
+    [Redirect-Server]
+      {Redirect-Address-Type}
+      {Redirect-Server-Address}
+```
+
+The support of Restriction-Filter-Rule can be added later if required. In this
+case, ReceivedCredits will need to be updated to include a FlowDescription list.
+
+We will need to translation IPFilterRule to FlowDescription as well.
+
+
+**SessionD**
+
+ServiceAction will need to be updated to support RESTRICT\_ACCESS action. Mostly,
+we will need to be able to store a restricted list of rule ids received from
+OCS/SessionProxy.
+
+In addition, we will need to store the list of rules ID to be used when restriction
+is enabled. A method get\_restricted\_rule\_ids will be added to retrive this list
+prior to sending rules activation to pipelined.
+
+We will re-use the same pipelined client method add\_gy\_final\_action\_flow as It
+provides support for this case as well.
+
+Static Rules will be sent using their original priority and will be applied as gy
+final action flow.
+
+The service state SERVICE\_RESTRICTED will need to be supported as well.
+
+
+**Pipelined**
+
+GYController needs to be updated to support restriction FUA. This translates to,
+1/ activate all the flows as they are received from SessionD.
+2/ Support QoS in GY APP (Include a Qos Manager)
+3/ Updating statistics back to SessionD
```

Comment:
I think if we are sending up stats for Gy rules, we will need some changes in SessionD as well to apply the Gy stats in. (Shouldn't be a big change, but we don't look at Gy rules when we currently take in enforcement stats)

Reply:
Yes, good point. I will add this to the list